### PR TITLE
[NOCHECKIN] Confidence Intervals using bootstrap

### PR DESCRIPTION
**Pull request by riedgar-ms**
## Description

This is an update of #1038 to sort out some git confusion. It will address #999 . The following description is taken from the original PR:

- Use of [bootstrap](https://en.wikipedia.org/wiki/Bootstrapping_(statistics)) to estimate the empirical distribution of any scalar metric passed to MetricFrame. Bootstrap settings are controlled by two new init parameters: `n_boot`, which determines how many replications of the dataset to generate with replacement, and `n_jobs`, which controls how many parallel processes to use in the bootstrap operation.

- Confidence interval estimation as a summary of the empirical bootstrap distribution. This is controlled by the `ci` and `ci_method` parameters, which offer options of which quantiles to sample and whether or not estimates should be bias-corrected. Many users can use simple parameter settings like `ci=0.95` for a 95% confidence interval on each MetricFrame statistic, while advanced users can query specific quantiles (e.g. `ci=[0.25, 0.5, 0.75]`) for further study. These results are calculated during initialization and are stored in new attributes on MetricFrame suffixed in `.*_ci`. 

Some significant changes have been made to MetricFrame internals in order to support this set of operations. Firstly, because we build the bootstrap distribution during initialization and chose not to pin it on the object, some operations like `.difference` are calculated during init. Secondly, because bootstrapping involves calculating the same operation on different (resampled) inputs, I've changed some internal functions like `def __group()` to work on any set of inputs, and not be hard-coded to use pinned attributes like `self.overall` and `self.by_group`. I've also refactored some general code into new private functions to make repeated calls easier. These should have no breaking changes to an end user. 

The new `.*_ci` attributes are all lists of tuples, with each tuple representing the quantile of the distribution in the first element, and the corresponding values for each metric in the second. For example, if a user requests a 90% confidence interval with `ci=0.9`, the output of `.overall_ci` will look like:

```python
.overall_ci = [
   (0.05, pd.DataFrame({...})),
   (0.95, pd.DataFrame({...}))
]
```

where each of the inner frames exactly parallel the structure of `.overall`. This API design hopefully makes it easy for users to adopt confidence intervals into their workflow -- any code that runs on the "main" MetricFrame objects will also work on the .ci versions.

## Tests

- [ ] no new tests required
- [x] new tests added
- [ ] existing tests adjusted

## Documentation

- [ ] no documentation changes needed
- [ ] user guide added or updated
- [x] API docs added or updated
- [ ] example notebook added or updated

### CODE REVIEW

File: fairlearn/metrics/_metric_frame.py [hunk 1]
```diff
@@ -364,6 +392,138 @@ def __init__(
...
+                    self.by_group, group_min_output
+                )
+                difference_overall_output = self.__difference(
+                    self.by_group, self.overall
+                )
+                ratio_group_output = group_min_output / group_max_output
+                ratio_overall_output = self._calc_ratio_overall(
+                    self.by_group, self.overall
+                )
+            except:
```
Comment (riedgar-ms) [inline fairlearn/metrics/_metric_frame.py:457]: Having been bitten by this when bring the PR over, I think it would be best not to completely throw out the actual exception. Given that the raised `ValueError` will likely be the most common for users, I can see the sense in keeping it, but I think we should nest the actual exception inside it.

File: test/unit/metrics/test_metricframe_bootstrap.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        mf.ratio(method="to_overall").shape == ci_output.shape
+        for _, ci_output in mf.ratio_overall_ci
+    )
+    assert all(
+        mf.ratio(method="between_groups").shape == ci_output.shape
+        for _, ci_output in mf.ratio_group_ci
+    )
+
+
+def check_all_bootstrap_directions(mf):
```
Comment (riedgar-ms) [inline test/unit/metrics/test_metricframe_bootstrap.py:58]: This should probably check that the values reported by `overall`, `difference()` etc. lie within the confidence intervals

File: fairlearn/metrics/_metric_frame.py [hunk 3]
```diff
@@ -651,19 +866,39 @@ def difference(
...
+                result = tmp.transform(lambda x: min(x, 1 / x)).min()
         else:
-            result = (mf - subtrahend).abs().groupby(level=self.control_levels).max()
+            if self.control_levels:
+                # It's easiest to give in to the DataFrame columns preference
+                ratios = self.by_group.unstack(
+                    level=self.control_levels
+                ) / self.overall.unstack(level=self.control_levels)
+            else:
+                ratios = self.by_group / self.overall
```
Comment (riedgar-ms) [inline fairlearn/metrics/_metric_frame.py:889]: Similarly here - and `overall_frame` too @harsha-nori ?

**Comment by riedgar-ms** [comment]
Also should add some API docs, and possibly an example notebook.

File: fairlearn/metrics/_metric_frame.py (hunk 1 above)
Comment (riedgar-ms) [reply fairlearn/metrics/_metric_frame.py]: Fixed this mone.

File: test/unit/metrics/test_metricframe_aggregates.py [hunk 4]
```diff
@@ -125,6 +125,17 @@ def test_ratio_to_overall(self, metric_fn):
             expected_ratio_overall, rel=1e-10, abs=1e-16
         )
 
+    def test_ratio_zero_divide(self):
```
Comment (riedgar-ms) [inline test/unit/metrics/test_metricframe_aggregates.py:128]: This is for a lurking bug... I can put into a separate PR if there are strong feelings, but it would be simpler to keep here.

File: fairlearn/metrics/_metric_frame.py [hunk 5]
```diff
@@ -651,19 +866,39 @@ def difference(
...
-        # Can assume errors='coerce', else error would already have been raised in .group_min
-        # Fill all non-scalar values with NaN
-        if isinstance(mf, pd.Series):
-            mf = mf.map(lambda x: x if np.isscalar(x) else np.nan)
-        else:
-            mf = mf.applymap(lambda x: x if np.isscalar(x) else np.nan)
+        return self.__difference(self.by_group, subtrahend)
+
+    def _calc_ratio_overall(self, by_group_frame, overall_frame):
+        def ratio_sub_one(x):
```
Comment (riedgar-ms) [inline fairlearn/metrics/_metric_frame.py:872]: This is part of the bugfix mentioned for the aggregates. Basically, this makes everything use `ratio_sub_one()` rather than using `min(x, 1/x)` in some places.

File: docs/user_guide/assessment.rst [hunk 6]
```diff
@@ -589,6 +589,94 @@ For more examples, please
...
+    [(0.025, 0.23333333333333334), (0.975, 0.6)]
+
+We have turned on bootstrapping by setting the :code:`n_boot` parameter;
+in this case, we have requested that 100 synthetic datasets be generated
+to create the statistics.
+We see that the :attr:`MetricFrame.overall` property is unchanged.
+However, we can also access the :attr:`MetricFrame.overall_ci` property
+which contains estimates of the confidence interval.
+By default, the bootstrapping algorithm computes the 95% confidence
+interval, which ranges from the 25th millile to the 975th millile.
```
Comment (riedgar-ms) [inline docs/user_guide/assessment.rst:647]: Feedback on whether 'millile' is the right word invited.

Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:651]: Or should I just go with 'quantile' and give decimals?

File: test/unit/metrics/test_metricframe_bootstrap_values.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+import math
+
+import numpy as np
+import pytest
+
+from sklearn.metrics import accuracy_score
+
+from fairlearn.metrics import MetricFrame
+
+# The setup is slightly unorthodx, to allow for splitting of
```
Comment (riedgar-ms) [inline test/unit/metrics/test_metricframe_bootstrap_values.py:19]: Comments welcome on this pattern, since I've not used it before. I'm trying to avoid having a huge test case with lots of assertions in it, but at the same time, having a bunch of stuff critical to the test cases defined at the top of the file like this isn't ideal either.

File: test/unit/metrics/test_metricframe_bootstrap_values.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        for i in range(len(sigma_vals)):
+            v_A = mean_A + (sigma_vals[i] * sigma_A)
+            vf_A = v_A / n_A
+
+            v_B = mean_B + (sigma_vals[i] * sigma_B)
+            vf_B = v_B / n_B
+
+            # Slightly wider tolerance here.... not sure if this should be required
+            assert mf.by_group_ci[i][1]["A"] == pytest.approx(
+                vf_A, rel=2 / math.sqrt(n_A)
```
Comment (riedgar-ms) [inline test/unit/metrics/test_metricframe_bootstrap_values.py:144]: @MiroDudik and @Harsha-Nori, your feedback here would be much appreciated. This is a wider tolerance than I was expecting to need, so I'm worried that I'm doing something wrong in the calculation.

File: test/unit/metrics/test_metricframe_bootstrap_values.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        # Do the error calculation
+        nominal_acc_AB = abs(p_A - p_B)
+
+        sigma_A = math.sqrt(n_A * p_A * (1 - p_A))
+        sigma_B = math.sqrt(n_B * p_B * (1 - p_B))
+        acc_sigma_A = sigma_A / n_A
+        acc_sigma_B = sigma_B / n_B
+
+        # The actual quadrature
+        acc_sigma_AB = math.sqrt(acc_sigma_A**2 + acc_sigma_B**2)
```
Comment (riedgar-ms) [inline test/unit/metrics/test_metricframe_bootstrap_values.py:169]: Again @MiroDudik  and @harsha-nori, I'd appreciate it if you double check my work here. It's been a while since I did error calculations.

**Comment by riedgar-ms** [comment]
Trying to add some checks on the actual values produced by the bootstrapping. I'd be very grateful if my approach could be double checked (before I go and fill out ratios etc.). @MiroDudik  and @harsha-nori especially.

**Comment by riedgar-ms** [comment]
Ping @MiroDudik  and @harsha-nori

**Comment by riedgar-ms** [comment]
Ping again.... @MiroDudik @harsha-nori

**Comment by riedgar-ms** [comment]
Ping @harsha-nori and @MiroDudik . I'd really appreciate some feedback on this before I expand the testing

File: fairlearn/metrics/_metric_frame.py [hunk 10]
```diff
@@ -309,6 +333,10 @@ def __init__(
         sample_params: Optional[
             Union[Dict[str, Any], Dict[str, Dict[str, Any]]]
         ] = None,
+        n_boot: Optional[int] = None,
+        ci: Union[float, List[float], Tuple[float]] = 0.95,
```
Comment (MiroDudik) [inline fairlearn/metrics/_metric_frame.py:337]: I think that the default should be [0.025, 0.975]

**Comment by riedgar-ms** [comment]
Spoke briefly with @MiroDudik about this yesterday. He suggests that not returning tuples, but instead:

- Having a `MetricFrame.quantiles_` property which stores the list of quantiles computed
- Convert the existing lists of Tuples into simple lists. These then are 'indexed' by `quantiles_`

I quite like this idea (stop returning lots of redundant data), but it will be a big change. So before I do so, I'm soliciting thoughts @Harsha-Nori, @adrinjalali @hildeweerts ?

File: docs/user_guide/assessment.rst (hunk 6 above)
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:877]: Will do. Although this particular part of the PR is not going to survive the other changes I'm making to the MetricFrame documentation.

File: fairlearn/metrics/_bootstrap.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+"""Utility functions for bootstrap calculations (used for confidence intervals)."""
+
+from warnings import warn
+import pandas as pd
+import numpy as np
+from scipy.stats import norm
+import numbers
+
+
+def process_ci_bounds(ci, ci_method, precision=6):
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply fairlearn/metrics/_bootstrap.py:13]: @harsha-nori can you comment? This bit was copied straight from your PR.

File: fairlearn/metrics/_metric_frame.py [hunk 12]
```diff
@@ -364,6 +392,138 @@ def __init__(
             all_data, annotated_funcs, grouping_features
         )
 
+        # Calculate uncertainty intervals using bootstrapped versions of data
+        def _bootstrap_metrics_calc(data):
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply fairlearn/metrics/_metric_frame.py:396]: Let me think about some possibilities. This was just a direct copy of @harsha-nori 's PR, to get it moving again. I'm certainly not a huge fan of how things are calculated right now.

File: fairlearn/metrics/_metric_frame.py [hunk 13]
```diff
@@ -405,6 +565,29 @@ def overall(self) -> Union[Any, pd.Series, pd.DataFrame]:
         else:
             return self._overall
 
+    @property
+    def overall_ci(
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply fairlearn/metrics/_metric_frame.py:583]: Have been applying this change today.

File: test/unit/metrics/test_metricframe_bootstrap_values.py (hunk 8 above)
Comment (riedgar-ms) [reply test/unit/metrics/test_metricframe_bootstrap_values.py]: Hmm.... this has been lost from the 'Files' tab unfortunately :-/


I think I get what you're saying here @MiroDudik , but I'd like to be certain.... Basically, what I was aiming to do here was verify that sensible values were being produced, because nothing in the existing tests did that. What you're saying @MiroDudik  is that the way I've gone about this is also testing whether or not bootstrapping works statistically (because I'm relying on certain properties of binomial and normal distributions, and then hammering down uncertainties with a large n_boot)? So a better approach would be to create a 'known' set of bootstrap samples, and then perform the quantile calculations on those?

File: test/unit/metrics/test_metricframe_bootstrap_values.py (hunk 9 above)
Comment (riedgar-ms) [reply test/unit/metrics/test_metricframe_bootstrap_values.py:176]: It sounds like you're suggesting testing the guts of `create_ci_output()` in `_bootstrap.py` - the stuff around `np.quantile()` specifically. I would actually be happy to do that; this was an option because I understood what was supposed to be going on at this high level. The expected output of `create_ci_output()` was not so easy to discern...

File: fairlearn/metrics/_metric_frame.py [hunk 14]
```diff
@@ -223,6 +226,27 @@ class MetricFrame:
...
+        not supported on complex metrics that yield multiple outputs.
+    ci : float, list, tuple
+        Determines the quantiles reported for confidence intervals. All entries must
+        be in the range [0, 1]. A single float is interpreted as the width of the
+        desired confidence interval (e.g. 0.95 is a 95% interval), while a List or Tuple
+        reports exact quantiles from the empirical bootstrap distribution
+        (e.g. (0, 0.33, 0.5, 0.66, 1) will calculate the min, 33rd, 50th, 66th, and max quantiles).
+    ci_method: str
+        One of {'percentile', 'bias-corrected'}. This method determines the algorithm used to
+        produce valid confidence intervals from the bootstrap distribution. 'percentine' returns
```
Comment (hildeweerts) [inline fairlearn/metrics/_metric_frame.py:243]:
```suggestion
        produce valid confidence intervals from the bootstrap distribution. 'percentile' returns
```

File: fairlearn/metrics/_metric_frame.py [hunk 15]
```diff
@@ -405,6 +565,29 @@ def overall(self) -> Union[Any, pd.Series, pd.DataFrame]:
...
+                    ]
+                else:
+                    return [
+                        (quantile, output.iloc[0])
+                        for quantile, output in self._overall_ci
+                    ]
+            else:
+                return self._overall_ci
+
+        return None
```
[the remark this answers is not in the document]
Comment (hildeweerts) [reply fairlearn/metrics/_metric_frame.py:597]: To me `None` makes sense, because that it throws an error if you try to iterate over `_overall_ci` which might be useful for debugging purposes. But I don't know what the best practices are :)

File: docs/user_guide/assessment.rst [hunk 16]
```diff
@@ -815,6 +815,120 @@ For more examples, please
...
+
+.. doctest:: assessment_metrics
+    :options:  +NORMALIZE_WHITESPACE
+
+    >>> metric_ci.difference(method='between_groups')
+    0.325
+    >>> metric_ci.difference_group_ci
+    [(0.025, 0.10214826839826842), (0.975, 0.7585227272727271)]
+
+What we see here is that the largest difference between any two of the groups
```
Comment (hildeweerts) [inline docs/user_guide/assessment.rst:925]: Should we also mention somewhere that fairness metrics such as max-min difference are (upwards) statistically biased estimates (see Section 4 / Figure 2 of  [Lum et al. 2022](https://arxiv.org/pdf/2205.05770.pdf)), or would that make things too complicated for our user guide?

**Review by hildeweerts — COMMENTED** [review]
Just a few small comments :)

**Comment by riedgar-ms** [comment]
Will be closing this in favour of #1125 . Keeping until I've retrieved the docs

**Comment by riedgar-ms** [comment]
Superseded
